File: src/Functional/Erlang/adapter.erl
-module(adapter).
-export([main/0]).

legacy_read_fahrenheit() -> 86.

adapt_to_celsius(ReadFahrenheit) ->
    fun() -> ((ReadFahrenheit() - 32) * 5) div 9 end.

main() ->
    ReadCelsius = adapt_to_celsius(fun legacy_read_fahrenheit/0),
    io:format("legacy=~BF~n", [legacy_read_fahrenheit()]),
    io:format("adapted=~BC~n", [ReadCelsius()]).

File: src/Functional/OCaml/adapter.ml
let legacy_read_fahrenheit () = 86

let adapt_to_celsius read_fahrenheit () =
  ((read_fahrenheit () - 32) * 5) / 9

let () =
  let read_celsius = adapt_to_celsius legacy_read_fahrenheit in
  Printf.printf "legacy=%dF\n" (legacy_read_fahrenheit ());
  Printf.printf "adapted=%dC\n" (read_celsius ())
